pkg/daemon/ceph/osd/remove.go

```diff
@@ -0,0 +1,124 @@
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// RemoveOSDs purges a list of OSDs from the cluster
+func RemoveOSDs(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdsToRemove []string) error {
+
+	// Generate the ceph config for running ceph commands similar to the operator
+	if err := writeCephConfig(context, clusterInfo); err != nil {
+		return errors.Wrap(err, "failed to write the ceph config")
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		logger.Infof("removing OSD %q", osdIDStr)
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
+			logger.Errorf("failed to remove OSD %d. %v", osdID, err)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
+
+	// TODO: Get the CRUSH map to determine
+	// 1. The host where the OSD is found
+	// 2. If the OSD is down
+	hostName := "minikube"
+	osdDown := true
+
+	if !osdDown {
+		return fmt.Errorf("OSD %d cannot be removed unless it is 'down'", osdID)
+	}
+
+	// Mark the OSD as out. If it is not down, it is expected this command will fail, in which case
+	// we expect to abort purging the OSD
+	output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("cannot remove the osd. %s", output))
+	}
+
+	// Remove the OSD deployment
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	logger.Infof("removing the OSD deployment %q", deploymentName)
+	if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+		if !kerrors.IsNotFound(err) {
+			// Continue purging the OSD even if the deployment fails to be deleted
+			logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+		}
+	}
+
+	// If running on a PVC, delete the OSD prepare job
+	prepareJobName := fmt.Sprintf("rook-ceph-osd-prepare-example-deviceset-0-#{osdID}")
+	prepareJob, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).Get(prepareJobName, metav1.GetOptions{})
+	logger.Infof("removing the OSD prepare job %q", prepareJobName)
+	if err := k8sutil.DeleteBatchJob(context.Clientset, clusterInfo.Namespace, prepareJobName, false); err != nil {
+		if !kerrors.IsNotFound(err) {
+			// Continue deleting the OSD prepare job even if the deployment fails to be deleted
+			logger.Errorf("failed to delete prepare job for OSD %s. %v", prepareJobName, err)
+		}
+	}
+
+	// If running on a PVC, delete the OSD PVC
+	if pvcName, ispresent := prepareJob.GetLabels()["ceph.rook.io/pvc"]; ispresent {
```

Comment:
in golang it's common to use `ok` for this check in the map
```suggestion
	if pvcName, ok := prepareJob.GetLabels()["ceph.rook.io/pvc"]; ok {
```

Reply:
The `prepareJob` might not have been found on line 87, so we need to skip this check if it's `nil`.